modules/engine/src/lib/program-manager.d.ts

```diff
@@ -0,0 +1,66 @@
+import {
+  VertexShader,
+  FragmentShader,
+  Program
+} from '@luma.gl/webgl';
+
+import {
+  Uniforms,
+  UniformsOptions
+} from '@luma.gl/webgl/src/classes/uniforms'
+
+import {
+  ShaderModule,
+  ShaderModuleObject,
+  InjectionMap
+} from '@luma.gl/shadertools/src/lib/shader-module'
+
+import {
+  DefineMap
+} from '@luma.gl/shadertools/src/lib/assemble-shaders'
+
+interface ProgramManagerProps {
+  vs?: VertexShader | string
+  fs?: FragmentShader | string
+  modules?: Array<ShaderModuleObject | string>
+  defines?: DefineMap
+  inject?: InjectionMap
+  varyings?: Array<string>
+  bufferMode?: number
+  transpileToGLSL100?: boolean
+}
+
+export default class ProgramManager {
+  stateHash: number
+  gl: WebGLRenderingContext
+  _programCache: {
+    [programHash: string]: Program
+  }
+  _getUniforms: {
+    [programHash: string]: (opts: UniformsOptions, context: Uniforms) => Uniforms
+  }
+  _registeredModules: {} // TODO: Remove? This isn't used anywhere in luma.gl
+  _hookFunctions: Array<{
+    hook: string
+    header?: string
+    footer?: string
+  }>
+  _defaultModules: Array<ShaderModuleObject | ShaderModule>
+  _hashes: {
+    [key: string]: number
+  }
+  _hashCounter: number
+  _useCounts: {
+    [programHash: string]: number
+  }
+  static getDefaultProgramManager(gl: WebGLRenderingContext): ProgramManager;
+  constructor(gl: WebGLRenderingContext);
+  addDefaultModule(module: ShaderModuleObject | ShaderModule): void;
+  removeDefaultModule(module: ShaderModuleObject | ShaderModule | string): void;
+  addShaderHook(hook: string, opts?: {header?: string, footer?: string}): void;
+  get(props?: ProgramManagerProps): Program;
+  getUniforms(program: Program): ((opts: UniformsOptions, context: Uniforms) => Uniforms) | null;
+  release(program: Program): void;
+  _getHash(key: string): number;
```

Comment:
Normally I don't add types for the private methods. Just JSDoc them inside the JS file

Reply:
I found it really helpful to add private typings here while working out the types (many private types infer the public types). I can move the private stuff to JSDoc, no worries.